File: apps/web/src/api/market.ts
import type { JobPosting } from '@job-copilot/shared/jobs';

import { readStoredProfileId, request } from './client';
import type {
  EngineMarketCompaniesResponse,
  EngineMarketCompanyDetail,
  EngineMarketCompanyVelocityEntry,
  EngineMarketCompanyVelocityTrend,
  EngineMarketFreezeSignalEntry,
  EngineMarketOverview,
  EngineMarketRegionDemandEntry,
  EngineMarketRoleDemandEntry,
  EngineMarketSalaryBySeniorityEntry,
  EngineMarketSalaryTrend,
  EngineMarketTechDemandEntry,
} from './engine-types';
import { companySlug } from '../lib/companySlug';
import { mapJob } from './mappers';

export interface SkillStat {
  skill: string;
  count: number;
  pct: number;
  inResume: boolean;
}

export interface MarketInsights {
  totalJobs: number;
  coverageScore: number;
  topSkills: SkillStat[];
  hotGaps: string[];
  salaryMentions: string[];
}

export type MarketTrend = 'up' | 'down' | 'stable';
export type MarketCompanyVelocityTrend = EngineMarketCompanyVelocityTrend;

export type MarketOverview = {
  newJobsThisWeek: number;
  activeCompaniesCount: number;
  activeJobsCount: number;
  remotePercentage: number;
};

export type MarketCompany = {
  companyName: string;
  normalizedCompanyName: string;
  activeJobs: number;
  thisWeek: number;
  prevWeek: number;
  velocity: number;
  sources: string[];
  topRoleGroups: string[];
  latestJobIds: string[];
  dataQualityFlags: string[];
};

export type MarketCompanyDetail = {
  companyName: string;
  normalizedCompanyName: string;
  totalJobs: number;
  activeJobs: number;
  avgSalary?: number;
  companyStatus?: 'whitelist' | 'blacklist';
  velocity: Array<{
    date: string;
    jobCount: number;
  }>;
  jobs: JobPosting[];
};

export type MarketCompanyVelocity = {
  company: string;
  jobCount: number;
  trend: MarketCompanyVelocityTrend;
};

export type MarketFreezeSignal = {
  company: string;
  lastPostedAt: string;
  daysSinceLastPost: number;
  historicalCount: number;
};

export type MarketSalaryTrend = {
  seniority: string;
  currency: string;
  p25: number;
  median: number;
  p75: number;
  sampleCount: number;
};

export type MarketSalaryBySeniority = {
  seniority: string;
  medianMin: number;
  medianMax: number;
  sampleSize: number;
};

export type MarketRoleDemand = {
  roleGroup: string;
  thisPeriod: number;
  prevPeriod: number;
  trend: MarketTrend;
};

export type MarketRegionDemand = {
  region: string;
  jobCount: number;
  topRoles: string[];
};

export type MarketTechDemand = {
  skill: string;
  jobCount: number;
  percentage: number;
};

const DEFAULT_SENIORITY_BUCKETS = ['junior', 'middle', 'senior', 'lead'] as const;

export { companySlug };

export async function getMarketOverview(): Promise<MarketOverview> {
  const response = await request<EngineMarketOverview>('/api/v1/market/overview');

  return {
    newJobsThisWeek: response.new_jobs_this_week,
    activeCompaniesCount: response.active_companies_count,
    activeJobsCount: response.active_jobs_count,
    remotePercentage: response.remote_percentage,
  };
}

export async function getMarketCompanies(limit = 10): Promise<MarketCompany[]> {
  const response = await request<EngineMarketCompaniesResponse>(
    `/api/v1/market/companies?limit=${encodeURIComponent(String(limit))}`,
  );

  return response.companies.map((company) => ({
    companyName: company.company_name,
    normalizedCompanyName:
      company.normalized_company_name ?? company.company_name.trim().toLowerCase(),
    activeJobs: company.active_jobs,
    thisWeek: company.this_week,
    prevWeek: company.prev_week,
    velocity: company.velocity,
    sources: company.sources ?? [],
    topRoleGroups: company.top_role_groups ?? [],
    latestJobIds: company.latest_job_ids ?? [],
    dataQualityFlags: company.data_quality_flags ?? [],
  }));
}

export async function getMarketCompanyDetail(slug: string): Promise<MarketCompanyDetail> {
  const params = new URLSearchParams();
  const profileId = readStoredProfileId();
  if (profileId) {
    params.set('profile_id', profileId);
  }
  const query = params.toString();
  const response = await request<EngineMarketCompanyDetail>(
    `/api/v1/market/companies/${encodeURIComponent(slug)}${query ? `?${query}` : ''}`,
  );

  return {
    companyName: response.company_name,
    normalizedCompanyName: response.normalized_company_name,
    totalJobs: response.total_jobs,
    activeJobs: response.active_jobs,
    avgSalary: response.avg_salary ?? undefined,
    companyStatus: response.company_status ?? undefined,
    velocity: response.velocity.map((point) => ({
      date: point.date,
      jobCount: point.job_count,
    })),
    jobs: response.jobs.map(mapJob),
  };
}

export async function getMarketCompanyVelocity(): Promise<MarketCompanyVelocity[]> {
  const response = await request<EngineMarketCompanyVelocityEntry[]>(
    '/api/v1/market/company-velocity',
  );

  return response.map((entry) => ({
    company: entry.company,
    jobCount: entry.job_count,
    trend: entry.trend,
  }));
}

export async function getMarketFreezeSignals(): Promise<MarketFreezeSignal[]> {
  const response = await request<EngineMarketFreezeSignalEntry[]>('/api/v1/market/freeze-signals');

  return response.map((entry) => ({
    company: entry.company,
    lastPostedAt: entry.last_posted_at,
    daysSinceLastPost: entry.days_since_last_post,
    historicalCount: entry.historical_count,
  }));
}

export async function getMarketSalaries(
  seniorityBuckets: readonly string[] = DEFAULT_SENIORITY_BUCKETS,
): Promise<MarketSalaryTrend[]> {
  const buckets = Array.from(
    new Set(seniorityBuckets.map((bucket) => bucket.trim().toLowerCase()).filter(Boolean)),
  );

  const response = await request<EngineMarketSalaryTrend[]>('/api/v1/market/salary-trends');

  const trendsBySeniority = new Map(
    response.map((trend) => [
      trend.seniority.toLowerCase(),
      {
        seniority: trend.seniority,
        currency: trend.currency,
        p25: trend.p25,
        median: trend.median,
        p75: trend.p75,
        sampleCount: trend.sample_count,
      } satisfies MarketSalaryTrend,
    ]),
  );

  return buckets
    .map((bucket) => trendsBySeniority.get(bucket))
    .filter((result): result is MarketSalaryTrend => result !== undefined);
}

export async function getMarketSalaryBySeniority(): Promise<MarketSalaryBySeniority[]> {
  const response = await request<EngineMarketSalaryBySeniorityEntry[]>(
    '/api/v1/market/salary-by-seniority',
  );

  return response.map((entry) => ({
    seniority: entry.seniority,
    medianMin: entry.median_min,
    medianMax: entry.median_max,
    sampleSize: entry.sample_size,
  }));
}

export async function getMarketRoles(period = 30): Promise<MarketRoleDemand[]> {
  const response = await request<EngineMarketRoleDemandEntry[]>(
    `/api/v1/market/roles?period=${encodeURIComponent(String(period))}`,
  );

  return response.map((entry) => ({
    roleGroup: entry.role_group,
    thisPeriod: entry.this_period,
    prevPeriod: entry.prev_period,
    trend: entry.trend,
  }));
}

export async function getMarketRegionBreakdown(): Promise<MarketRegionDemand[]> {
  const response = await request<EngineMarketRegionDemandEntry[]>(
    '/api/v1/market/region-breakdown',
  );

  return response.map((entry) => ({
    region: entry.region,
    jobCount: entry.job_count,
    topRoles: entry.top_roles ?? [],
  }));
}

export async function getMarketTechDemand(): Promise<MarketTechDemand[]> {
  const response = await request<EngineMarketTechDemandEntry[]>('/api/v1/market/tech-demand');

  return response.map((entry) => ({
    skill: entry.skill,
    jobCount: entry.job_count,
    percentage: entry.percentage,
  }));
}

File: apps/engine-api/src/api/dto/market.rs
use serde::Serialize;

use crate::api::dto::jobs::JobResponse;
use crate::domain::feedback::model::JobFeedbackState;
use crate::domain::market::model::{
    MarketCompanyDetail, MarketCompanyEntry, MarketCompanyVelocityEntry,
    MarketCompanyVelocityPoint, MarketFreezeSignalEntry, MarketOverview, MarketRegionDemandEntry,
    MarketRoleDemandEntry, MarketSalaryBySeniorityEntry, MarketSalaryTrend, MarketTechDemandEntry,
};

#[derive(Debug, Serialize)]
pub struct MarketOverviewResponse {
    pub new_jobs_this_week: i64,
    pub active_companies_count: i64,
    pub active_jobs_count: i64,
    pub remote_percentage: f64,
}

impl From<MarketOverview> for MarketOverviewResponse {
    fn from(m: MarketOverview) -> Self {
        Self {
            new_jobs_this_week: m.new_jobs_this_week,
            active_companies_count: m.active_companies_count,
            active_jobs_count: m.active_jobs_count,
            remote_percentage: m.remote_percentage,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketCompanyEntryResponse {
    pub company_name: String,
    pub normalized_company_name: String,
    pub active_jobs: i64,
    pub this_week: i64,
    pub prev_week: i64,
    pub velocity: i64,
    pub sources: Vec<String>,
    pub top_role_groups: Vec<String>,
    pub latest_job_ids: Vec<String>,
    pub data_quality_flags: Vec<String>,
}

impl From<MarketCompanyEntry> for MarketCompanyEntryResponse {
    fn from(e: MarketCompanyEntry) -> Self {
        Self {
            velocity: e.this_week - e.prev_week,
            company_name: e.company_name,
            normalized_company_name: e.normalized_company_name,
            active_jobs: e.active_jobs,
            this_week: e.this_week,
            prev_week: e.prev_week,
            sources: e.sources,
            top_role_groups: e.top_role_groups,
            latest_job_ids: e.latest_job_ids,
            data_quality_flags: e.data_quality_flags,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketCompaniesResponse {
    pub companies: Vec<MarketCompanyEntryResponse>,
}

#[derive(Debug, Serialize)]
pub struct MarketCompanyVelocityPointResponse {
    pub date: String,
    pub job_count: i64,
}

impl From<MarketCompanyVelocityPoint> for MarketCompanyVelocityPointResponse {
    fn from(value: MarketCompanyVelocityPoint) -> Self {
        Self {
            date: value.date,
            job_count: value.job_count,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketCompanyDetailResponse {
    pub company_name: String,
    pub normalized_company_name: String,
    pub total_jobs: i64,
    pub active_jobs: i64,
    pub avg_salary: Option<i32>,
    pub velocity: Vec<MarketCompanyVelocityPointResponse>,
    pub company_status: Option<String>,
    pub jobs: Vec<JobResponse>,
}

impl MarketCompanyDetailResponse {
    pub fn from_detail(
        detail: MarketCompanyDetail,
        company_status: Option<String>,
        feedback_states: Vec<JobFeedbackState>,
    ) -> Self {
        Self {
            company_name: detail.company_name,
            normalized_company_name: detail.normalized_company_name,
            total_jobs: detail.total_jobs,
            active_jobs: detail.active_jobs,
            avg_salary: detail.avg_salary,
            velocity: detail
                .velocity
                .into_iter()
                .map(MarketCompanyVelocityPointResponse::from)
                .collect(),
            company_status,
            jobs: detail
                .active_job_views
                .into_iter()
                .zip(feedback_states)
                .map(|(job, feedback)| JobResponse::from_view_with_feedback(job, feedback))
                .collect(),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketCompanyVelocityEntryResponse {
    pub company: String,
    pub job_count: i64,
    pub trend: String,
}

impl From<MarketCompanyVelocityEntry> for MarketCompanyVelocityEntryResponse {
    fn from(value: MarketCompanyVelocityEntry) -> Self {
        Self {
            company: value.company,
            job_count: value.job_count,
            trend: value.trend.as_str().to_string(),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketFreezeSignalEntryResponse {
    pub company: String,
    pub last_posted_at: String,
    pub days_since_last_post: u32,
    pub historical_count: u32,
}

impl From<MarketFreezeSignalEntry> for MarketFreezeSignalEntryResponse {
    fn from(value: MarketFreezeSignalEntry) -> Self {
        Self {
            company: value.company,
            last_posted_at: value.last_posted_at,
            days_since_last_post: value.days_since_last_post,
            historical_count: value.historical_count,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketSalaryTrendResponse {
    pub seniority: String,
    pub currency: String,
    pub p25: i32,
    pub median: i32,
    pub p75: i32,
    pub sample_count: i64,
}

impl From<MarketSalaryTrend> for MarketSalaryTrendResponse {
    fn from(value: MarketSalaryTrend) -> Self {
        Self {
            seniority: value.seniority,
            currency: value.currency,
            p25: value.p25,
            median: value.median,
            p75: value.p75,
            sample_count: value.sample_count,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketSalaryBySeniorityEntryResponse {
    pub seniority: String,
    pub median_min: u32,
    pub median_max: u32,
    pub sample_size: u32,
}

impl From<MarketSalaryBySeniorityEntry> for MarketSalaryBySeniorityEntryResponse {
    fn from(value: MarketSalaryBySeniorityEntry) -> Self {
        Self {
            seniority: value.seniority,
            median_min: value.median_min,
            median_max: value.median_max,
            sample_size: value.sample_size,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketRoleDemandEntryResponse {
    pub role_group: String,
    pub this_period: i64,
    pub prev_period: i64,
    pub trend: String,
}

impl From<MarketRoleDemandEntry> for MarketRoleDemandEntryResponse {
    fn from(value: MarketRoleDemandEntry) -> Self {
        Self {
            role_group: value.role_group,
            this_period: value.this_period,
            prev_period: value.prev_period,
            trend: value.trend.as_str().to_string(),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketRegionDemandEntryResponse {
    pub region: String,
    pub job_count: u32,
    pub top_roles: Vec<String>,
}

impl From<MarketRegionDemandEntry> for MarketRegionDemandEntryResponse {
    fn from(value: MarketRegionDemandEntry) -> Self {
        Self {
            region: value.region,
            job_count: value.job_count,
            top_roles: value.top_roles,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MarketTechDemandEntryResponse {
    pub skill: String,
    pub job_count: u32,
    pub percentage: f32,
}

impl From<MarketTechDemandEntry> for MarketTechDemandEntryResponse {
    fn from(value: MarketTechDemandEntry) -> Self {
        Self {
            skill: value.skill,
            job_count: value.job_count,
            percentage: value.percentage,
        }
    }
}
